system/system.go

```diff
@@ -47,13 +61,185 @@ func Info() (types.System, error) {
 		system.Network = network
 	}
 
+	vmSystem, vmRole, err := host.Virtualization()
+	if err == nil {
+		system.VMSystem = vmSystem
+		system.VMRole = vmRole
+	}
+
+	if runtime.GOOS == "linux" {
+		libcType, err := getLibCType()
+		if err == nil {
+			system.LibCType = libcType
+		}
+	}
+
 	return system, nil
 }
 
+func getLibCType() (string, error) {
+	output, err := exec.Command("ldd", "--version").CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+	text := strings.ToLower(string(output))
+	if strings.Contains(text, "gnu") {
+		return "glibc", nil
+	}
+	if strings.Contains(text, "musl") {
+		return "musl", nil
+	}
+	return "", nil
+}
+
+// GetCloudProvider inspects local files, looks up hostnames, and makes HTTP
+// requests in order to determine if the local system is running within a cloud
+// provider.
+func GetCloudProvider(ctx context.Context) string {
+	switch runtime.GOOS {
+	case "linux":
+		// Issue a dmidecode command to see if we are on EC2 or GCP
+		logger.Debug("sudo -n dmidecode -s bios-version")
+		output, err := exec.Command("sudo", "-n", "dmidecode", "-s", "bios-version").CombinedOutput()
+		if err != nil {
+			logger.WithError(err).Debug("couldn't run dmidecode")
+			return cloudMetadataFallback(ctx)
+		}
+		text := strings.ToLower(string(output))
+		if strings.Contains(text, "amazon") {
+			logger.Debug("Running on EC2")
+			return "EC2"
+		}
+		if strings.Contains(text, "google") {
+			logger.Debug("Running on GCP")
+			return "GCP"
+		}
+		// At this point, we need to issue a slightly different command to see
+		// if we are on Azure
+		logger.Debug("sudo -n dmidecode -s system-manufacturer")
+		output, err = exec.Command("sudo", "-n", "dmidecode", "-s", "system-manufacturer").CombinedOutput()
+		if err != nil {
+			logger.WithError(err).Debug("couldn't run dmidecode")
+			return cloudMetadataFallback(ctx)
+		}
+		text = strings.ToLower(string(output))
+		if strings.Contains(text, "microsoft") {
+			logger.Debug("Running on Azure")
+			return "Azure"
+		}
+	case "windows":
+		return cloudMetadataFallback(ctx)
```

Comment:
We can't detect cloud providers on windows systems? Is this a todo or are we just not gonna support it?

Reply:
cloudMetadataFallback does detect cloud providers, it just uses a more expensive method than running dmidecode or inspecting a file.

Reply:
Ah so the respective `dial...` functions should be able to retrieve windows info. Sounds good!